spring-cloud-gcp-data-datastore/src/main/java/org/springframework/cloud/gcp/data/datastore/core/convert/TwoStepsConversions.java

```diff
@@ -135,43 +163,74 @@ public TwoStepsConversions(CustomConversions customConversions, ObjectToKeyFacto
+					(BaseEntity) value);
+		}
+		throw new DatastoreDataException(
+				"Embedded entity was expected, but " + value.getClass() + " found");
 	}
 
 	@SuppressWarnings("unchecked")
-	private <T> T convertOnReadSingleEmbedded(Object value, Class<?> targetClass) {
+	private <T> T convertOnReadSingleEmbedded(Object value,
+			TypeInformation<?> targetTypeInformation) {
 		Assert.notNull(value, "Cannot convert a null value.");
 		if (value instanceof BaseEntity) {
-			return (T) this.datastoreEntityConverter.read(targetClass, (BaseEntity) value);
+			return (T) this.datastoreEntityConverter.read(targetTypeInformation.getType(),
+					(BaseEntity) value);
 		}
-		throw new DatastoreDataException("Embedded entity was expected, but " + value.getClass() + " found");
+		throw new DatastoreDataException("Embedded entity was expected, but "
+				+ value.getClass() + " found");
 	}
 
 	@SuppressWarnings("unchecked")
-	private <T> T convertOnReadSingle(Object val, Class<?> targetType) {
+	private <T> T convertOnReadSingle(Object val,
+			TypeInformation<?> targetTypeInformation) {
+		Class targetType = targetTypeInformation.getType();
+		Class sourceType = val.getClass();
 		Assert.notNull(val, "Cannot convert a null value.");
 		Object result = null;
 		TypeTargets typeTargets = computeTypeTargets(targetType);
 
-		if (typeTargets.getFirstStepTarget() == null && typeTargets.getSecondStepTarget() == null
+		if (typeTargets.getFirstStepTarget() == null
+				&& typeTargets.getSecondStepTarget() == null
 				&& ClassUtils.isAssignable(targetType, val.getClass())) {
 			//neither first or second steps were applied, no conversion is necessary
 			result = val;
 		}
-		else if (typeTargets.getFirstStepTarget() == null && typeTargets.getSecondStepTarget() != null) {
+		else if (typeTargets.getFirstStepTarget() == null &&
+				typeTargets.getSecondStepTarget() != null) {
 			//only second step was applied on write
 			result = this.internalConversionService.convert(val, targetType);
 		}
-		else if (typeTargets.getFirstStepTarget() != null && typeTargets.getSecondStepTarget() == null) {
+		else if (typeTargets.getFirstStepTarget() != null &&
+				typeTargets.getSecondStepTarget() == null) {
 			//only first step was applied on write
 			result = this.conversionService.convert(val, targetType);
 		}
-		else if (typeTargets.getFirstStepTarget() != null && typeTargets.getSecondStepTarget() != null) {
+		else if (typeTargets.getFirstStepTarget() != null &&
+				typeTargets.getSecondStepTarget() != null) {
 			//both steps were applied
-			Object secondStepVal = this.internalConversionService.convert(val, typeTargets.getFirstStepTarget());
+			Object secondStepVal = this.internalConversionService.convert(val,
+					typeTargets.getFirstStepTarget());
 			result = this.conversionService.convert(secondStepVal, targetType);
 		}
+		// if the value can be directly converted
+		else if (DatastoreNativeTypes.isNativeType(sourceType)
```

Comment:
no need for that, computeTypeTargets and the previous conditions take care of that

Reply:
It does not. 

For example it will not convert Boolean to String because the source type is already a datastore native type. This came up in my unit test for an embedded map with a non-string key type.